Source/Core/VideoCommon/UberShaderPixel.cpp

```diff
@@ -0,0 +1,716 @@
+		// compute window position if needed because binding semantic WPOS is not widely supported
+		if (numTexgen > 0)
+			out.Write(",\n  in %s float3 tex[%d] : TEXCOORD0", GetInterpolationQualifier(ApiType), numTexgen);
+		out.Write(",\n  in %s float4 clipPos : TEXCOORD%d", GetInterpolationQualifier(ApiType), numTexgen);
+		if (g_ActiveConfig.bEnablePixelLighting)
+		{
+			out.Write(",\n  in %s float3 Normal : TEXCOORD%d", GetInterpolationQualifier(ApiType), numTexgen + 1);
+			out.Write(",\n  in %s float3 WorldPos : TEXCOORD%d", GetInterpolationQualifier(ApiType), numTexgen + 2);
+		}
+		if (g_ActiveConfig.iStereoMode > 0)
+			out.Write(",\n  in uint layer : SV_RenderTargetArrayIndex\n");
+		out.Write("        ) {\n");
+	}
+
+	out.Write(
+		"	int AlphaBump = 0;\n"
+		"	int4 icolors_0 = int4(colors_0 * 255.0);\n"
+		"	int4 icolors_1 = int4(colors_1 * 255.0);\n"
+		"	int4 TevResult;\n"
+		"	State s;\n"
+		"	s.TexColor = int4(0, 0, 0, 0);\n"
+		"	s.RasColor = int4(0, 0, 0, 0);\n"
+		"	s.KonstColor = int4(0, 0, 0, 0);\n"
+		"\n");
+	for (int i = 0; i < 4; i++) out.Write(
+		"	s.Reg[%d] = " I_COLORS"[%d];\n", i, i);
+
+
+	out.Write(
+		"\n"
+		"	uint num_stages = %s;\n\n", BitfieldExtract("bpmem_genmode", GenMode().numtevstages).c_str());
+
+	out.Write(
+		"	// Main tev loop\n");
+	if (ApiType == API_D3D) out.Write(
+		"	[loop]\n"); // Tell DirectX we don't want this loop unrolled (it crashes if it tries to)
+	out.Write(
+		"	for(uint stage = 0u; stage <= num_stages; stage++)\n"
+		"	{\n"
+		"		uint cc = bpmem_combiners[stage].x;\n"
+		"		uint ac = bpmem_combiners[stage].y;\n"
+		"		uint order = bpmem_tevorder[stage>>1];\n"
+		"		if ((stage & 1u) == 1u)\n"
+		"			order = order >> %d;\n\n", TwoTevStageOrders().enable1.offset - TwoTevStageOrders().enable0.offset);
+
+	// TODO: Indirect Texturing
+	out.Write("\t\t// TODO: Indirect textures\n\n");
+
+	// Disable texturing when there are no texgens (for now)
+	if (numTexgen != 0)
+	{
+	out.Write(
+		"		// Sample texture for stage\n"
+		"		int4 texColor;\n"
+		"		if((order & %du) != 0u) {\n", 1 << TwoTevStageOrders().enable0.offset);
+	out.Write(
+		"			// Texture is enabled\n"
+		"			uint sampler_num = %s;\n", BitfieldExtract("order", TwoTevStageOrders().texmap0).c_str());
+	out.Write(
+		"			uint tex_coord = %s;\n", BitfieldExtract("order", TwoTevStageOrders().texcoord0).c_str());
+	out.Write(
+		"\n"
+		"			// TODO: there is an optional perspective divide here (not to mention all of indirect)\n"
+		"			int2 fixedPoint_uv = itrunc(tex[tex_coord].xy * " I_TEXDIMS"[sampler_num].zw * 128.0);\n"
+		"			float2 uv = (float2(fixedPoint_uv) / 128.0) * " I_TEXDIMS"[sampler_num].xy;\n"
+		"\n"
+		"			int4 color = sampleTexture(sampler_num, uv);\n"
+		"\n"
+		"			uint swap = %s;\n", BitfieldExtract("ac", TevStageCombiner().alphaC.tswap).c_str());
+	out.Write(
+		//"			s.TexColor.r = Swap(0, color);\n");
+		"			s.TexColor.r = Swap(3u-%s, color);\n", BitfieldExtract("bpmem_tevksel[swap*2]", TevKSel().swap1).c_str());
```

Comment:
These should be Swap(%s, color) instead of swap(3u-%s, color)

The commented out swap are hard-coded for TexColor.rgba = color.rgba, which makes it behave the same as the previous commit.